File: frontend/src/app/shared/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { User, UserStats } from '../models/user.model';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiUrl = 'http://localhost:8000/api/users';

  constructor(private http: HttpClient) {}

  getAllUsers(): Observable<User[]> {
    return this.http.get<User[]>(this.apiUrl)
      .pipe(
        catchError(error => {
          console.error('Get all users error:', error);
          return throwError(() => error);
        })
      );
  }

  getUserById(id: string): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/${id}`)
      .pipe(
        catchError(error => {
          console.error(`Get user ${id} error:`, error);
          return throwError(() => error);
        })
      );
  }

  getUserStats(): Observable<UserStats> {
    return this.http.get<UserStats>(`${this.apiUrl}/stats`)
      .pipe(
        catchError(error => {
          console.error('Get user stats error:', error);
          return throwError(() => error);
        })
      );
  }

  createUser(user: any): Observable<User> {
    return this.http.post<User>(this.apiUrl, user)
      .pipe(
        catchError(error => {
          console.error('Create user error:', error);
          return throwError(() => error);
        })
      );
  }

  updateUser(id: string, user: any): Observable<User> {
    return this.http.put<User>(`${this.apiUrl}/${id}`, user)
      .pipe(
        catchError(error => {
          console.error(`Update user ${id} error:`, error);
          return throwError(() => error);
        })
      );
  }

  deleteUser(id: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${id}`)
      .pipe(
        catchError(error => {
          console.error(`Delete user ${id} error:`, error);
          return throwError(() => error);
        })
      );
  }

  resetPassword(id: string, newPassword: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/${id}/reset-password`, { password: newPassword })
      .pipe(
        catchError(error => {
          console.error(`Reset password for user ${id} error:`, error);
          return throwError(() => error);
        })
      );
  }

  toggleBiometric(id: string, enabled: boolean): Observable<User> {
    return this.http.post<User>(`${this.apiUrl}/${id}/toggle-biometric`, { enabled })
      .pipe(
        catchError(error => {
          console.error(`Toggle biometric for user ${id} error:`, error);
          return throwError(() => error);
        })
      );
  }
}

File: backend/src/main/java/com/auth/archetype/service/UserService.java
package com.auth.archetype.service;

import java.util.List;
import java.util.Map;

import com.auth.archetype.dto.UserDto;
import com.auth.archetype.model.User;

public interface UserService {
    
    /**
     * Get all users
     * 
     * @return list of all users
     */
    List<UserDto> getAllUsers();
    
    /**
     * Get user by ID
     * 
     * @param id user ID
     * @return user data
     */
    UserDto getUserById(String id);
    
    /**
     * Get user by username
     * 
     * @param username username
     * @return user data
     */
    User getUserByUsername(String username);
    
    /**
     * Create a new user
     * 
     * @param userDto user data
     * @return created user
     */
    UserDto createUser(UserDto userDto);
    
    /**
     * Update an existing user
     * 
     * @param id user ID
     * @param userDto updated user data
     * @return updated user
     */
    UserDto updateUser(String id, UserDto userDto);
    
    /**
     * Delete a user
     * 
     * @param id user ID
     */
    void deleteUser(String id);
    
    /**
     * Reset a user's password
     * 
     * @param id user ID
     * @param newPassword new password
     */
    void resetPassword(String id, String newPassword);
    
    /**
     * Toggle biometric authentication for a user
     * 
     * @param id user ID
     * @param enabled whether biometric authentication should be enabled
     * @return updated user
     */
    UserDto toggleBiometric(String id, boolean enabled);
    
    /**
     * Get user statistics
     * 
     * @return map of statistics
     */
    Map<String, Long> getUserStats();
    
    /**
     * Get current user's activity information
     * 
     * @return map of activity data
     */
    Map<String, Object> getUserActivity();
    
    /**
     * Save user changes to persistent storage
     * 
     * @param user user to save
     * @return saved user
     */
    User saveUser(User user);
}
